File: jarvis/app/modules/ute-ble/ios/src/UteBleModule.swift
import ExpoModulesCore

/// Expo drops the message of `promise.reject(code, message)` ("undefined reason"),
/// so every failure goes through this, which keeps it.
final class UteException: GenericException<String> {
  override var reason: String { param }
}

// Every vendor SDK call lives in UteBleBridge.m; this file only speaks Foundation types.
public class UteBleModule: Module {
  private let bridge = UteBleBridge.shared

  private var syncPromise: Promise?
  private var syncFileExtension = "opus"

  public func definition() -> ModuleDefinition {
    Name("UteBle")

    Events(
      "onDeviceFound",
      "onConnectionChange",
      "onPairingChange", // The post-connect handshake on both platforms.
      "onBluetoothState",
      "onRecordStart",
      "onRecordStop",
      "onSyncProgress",
      "onClipData",
      "onInput",
      "onMotion",
      "onLog"
    )

    // The SDK touches CoreBluetooth on setup, so it stays on main.
    AsyncFunction("initialize") { () -> String in
      self.bridge.onDeviceFound = { [weak self] device in
        self?.sendEvent("onDeviceFound", device)
      }
      self.bridge.onConnectionChange = { [weak self] status, connected, error in
        self?.sendEvent("onConnectionChange", [
          "status": status,
          "connected": connected,
          "error": error as Any,
        ])
      }
      self.bridge.onBluetoothState = { [weak self] state, poweredOn in
        self?.sendEvent("onBluetoothState", ["state": state, "poweredOn": poweredOn])
      }
      self.bridge.onPairing = { [weak self] paired, message in
        self?.sendEvent("onPairingChange", ["paired": paired, "message": message])
      }
      self.bridge.onLog = { [weak self] line in
        self?.sendEvent("onLog", ["message": line])
      }
      self.bridge.onInput = { [weak self] input in
        self?.sendEvent("onInput", input)
      }
      self.bridge.onMotion = { [weak self] source, samples in
        self?.sendEvent("onMotion", ["source": source, "samples": samples])
      }
      self.bridge.onClip = { [weak self] data in
        self?.sendEvent("onClipData", ["bytes": data.count, "base64": data.base64EncodedString()])
      }
      self.bridge.onRecordStart = { [weak self] info in
        self?.sendEvent("onRecordStart", info)
      }
      self.bridge.onRecordStop = { [weak self] info in
        self?.sendEvent("onRecordStop", info)
      }
      self.bridge.onSyncProgress = { [weak self] sessionId, received, total in
        self?.sendEvent("onSyncProgress", [
          "sessionId": sessionId,
          "received": received,
          "total": total,
          "completed": false,
        ])
      }
      self.bridge.onSyncFinished = { [weak self] sessionId, data, error in
        self?.finishSync(sessionId: sessionId, data: data, error: error)
      }
      return self.bridge.setUp()
    }
    .runOnQueue(.main)

    AsyncFunction("startScan") {
      self.bridge.startScan()
    }
    .runOnQueue(.main)

    AsyncFunction("stopScan") {
      self.bridge.stopScan()
    }
    .runOnQueue(.main)

    AsyncFunction("connect") { (id: String) in
      guard self.bridge.connect(deviceId: id) else {
        throw UteException("UnknownDevice: the clip \(id) hasn't been seen by a scan yet. Scan again.")
      }
    }
    .runOnQueue(.main)

    AsyncFunction("disconnect") { () -> Bool in
      self.bridge.disconnect()
    }
    .runOnQueue(.main)

    AsyncFunction("isConnected") { () -> Bool in
      self.bridge.isConnected()
    }

    /// hasAIRecording says whether this firmware speaks the recording protocol at all.
    AsyncFunction("connectedDevice") { () -> [String: Any]? in
      self.bridge.connectedDevice()
    }

    /// Every `has…` capability flag the connected clip reports.
    AsyncFunction("capabilities") { () -> [String: Bool] in
      self.bridge.capabilities().mapValues { $0.boolValue }
    }
    .runOnQueue(.main)

    AsyncFunction("bind") { (token: String, verify: Int, promise: Promise) in
      self.bridge.bind(token: token, verify: verify) { errorCode, result in
        Self.settle(promise, errorCode, result, "appBindRecordDevice")
      }
    }

    AsyncFunction("getStatus") { (promise: Promise) in
      self.bridge.getStatus { errorCode, result in
        Self.settle(promise, errorCode, result, "getRecordStatus")
      }
    }

    AsyncFunction("getStorageInfo") { (promise: Promise) in
      self.bridge.getStorageInfo { errorCode, result in
        Self.settle(promise, errorCode, result, "getStorageCapacityInfo")
      }
    }

    AsyncFunction("getBattery") { (promise: Promise) in
      self.bridge.getBattery { errorCode, result in
        Self.settle(promise, errorCode, result, "getBatteryInfo")
      }
    }

    AsyncFunction("getRssi") { (promise: Promise) in
      self.bridge.getRssi { errorCode, result in
        Self.settle(promise, errorCode, result, "readDeviceRSSI")
      }
    }
    .runOnQueue(.main)

    AsyncFunction("getEncodingConfig") { (promise: Promise) in
      self.bridge.getEncodingConfig { errorCode, result in
        Self.settle(promise, errorCode, result, "getRecordEncodingConfiguration")
      }
    }

    AsyncFunction("probeSensors") { (promise: Promise) in
      self.bridge.probeSensors { errorCode, result in
        Self.settle(promise, errorCode, result, "checkFactoryFuntion")
      }
    }
    .runOnQueue(.main)

    AsyncFunction("readGyro") { (promise: Promise) in
      self.bridge.readGyro { errorCode, result in
        Self.settle(promise, errorCode, result, "factoryReadGyroData")
      }
    }
    .runOnQueue(.main)

    AsyncFunction("setMotionStream") { (on: Bool, promise: Promise) in
      self.bridge.setMotionStream(on: on) { errorCode in
        if errorCode == 408 {
          promise.reject(UteException("MotionUnsupported: the clip didn't answer sendGameStatus; this firmware likely has no motion stream."))
          return
        }
        guard errorCode == 0 else {
          promise.reject(UteException("sendGameStatus failed: \(Self.describe(errorCode))."))
          return
        }
        promise.resolve(nil)
      }
    }
    .runOnQueue(.main)

    AsyncFunction("setMotionSource") { (source: String, on: Bool, intervalMs: Int, promise: Promise) in
      self.bridge.setMotionSource(source, on: on, intervalMs: intervalMs) { errorCode in
        if errorCode == 408 {
          promise.reject(UteException("MotionUnsupported: the clip didn't answer (\(source))."))
          return
        }
        guard errorCode == 0 else {
          promise.reject(UteException("Motion source \(source) failed: \(Self.describe(errorCode))."))
          return
        }
        promise.resolve(nil)
      }
    }
    .runOnQueue(.main)

    AsyncFunction("readActivity") { (promise: Promise) in
      self.bridge.readActivity { errorCode, result in
        Self.settle(promise, errorCode, result, "getCurrentDayTotalWorkoutData")
      }
    }
    .runOnQueue(.main)

    /// Forward every SDK log line (raw packets included) as onLog, not only during a connect.
    AsyncFunction("setSdkLogging") { (on: Bool) in
      self.bridge.sdkLogging = on
    }
    .runOnQueue(.main)

    AsyncFunction("probeWearFunctions") { (promise: Promise) in
      self.bridge.probeWearFunctions { errorCode, result in
        Self.settle(promise, errorCode, result, "factoryReadFunction")
      }
    }
    .runOnQueue(.main)

    AsyncFunction("buzz") { (count: Int, option: Int, promise: Promise) in
      self.bridge.buzz(count: count, option: option) { errorCode, result in
        Self.settle(promise, errorCode, result, "buzz option \(option)")
      }
    }
    .runOnQueue(.main)

    AsyncFunction("startRecord") { (promise: Promise) in
      self.bridge.startRecord { errorCode, result in
        Self.settle(promise, errorCode, result, "startRecord")
      }
    }

    AsyncFunction("pauseRecord") { (sessionId: Int, promise: Promise) in
      self.bridge.pauseRecord(sessionId: sessionId) { errorCode, result in
        Self.settle(promise, errorCode, result, "pauseRecord")
      }
    }

    AsyncFunction("resumeRecord") { (sessionId: Int, promise: Promise) in
      self.bridge.resumeRecord(sessionId: sessionId) { errorCode, result in
        Self.settle(promise, errorCode, result, "resumeRecord")
      }
    }

    AsyncFunction("stopRecord") { (promise: Promise) in
      self.bridge.stopRecord { errorCode, result in
        Self.settle(promise, errorCode, result, "stopRecord")
      }
    }

    AsyncFunction("listFiles") { (promise: Promise) in
      self.bridge.listFiles { errorCode, result in
        Self.settle(promise, errorCode, result, "getRecordFileList")
      }
    }

    /// Pulls one recording over BLE into Documents/recordings, decodes it to WAV,
    /// and resolves with both paths.
    AsyncFunction("syncFile") { (sessionId: Int, fileType: Int, size: Int, promise: Promise) in
      if self.syncPromise != nil {
        promise.reject(UteException("SyncBusy: another download from the clip is still running."))
        return
      }
      guard let ext = Self.fileExtension(for: fileType) else {
        promise.reject(UteException("BadFileType: unknown file type \(fileType)."))
        return
      }
      self.syncPromise = promise
      self.syncFileExtension = ext

      self.bridge.syncFile(sessionId: sessionId, fileType: fileType, size: size) { errorCode, result in
        // result != 0: 1 filesystem error, 2 missing, 3 interrupted. No data follows.
        guard errorCode == 0, result == 0 else {
          let why = [1: "the clip's file system failed", 2: "the file isn't on the clip", 3: "the file is damaged"][result]
          self.syncPromise?.reject(
            UteException("SyncRefused: \(why ?? "the clip refused the download") (error \(errorCode), result \(result)).")
          )
          self.syncPromise = nil
          return
        }
      }
    }
    .runOnQueue(.main)

    AsyncFunction("stopSync") { (promise: Promise) in
      self.bridge.stopSync { _ in
        promise.resolve(nil)
      }
    }
    .runOnQueue(.main)

    /// Decodes a raw clip file (already on the phone) to WAV next to it.
    AsyncFunction("decodeFile") { (path: String) -> [String: Any] in
      let source = URL(fileURLWithPath: path)
      let data = try Data(contentsOf: source)
      return try Self.decodeToWav(data, next: source)
    }

    AsyncFunction("deleteFile") { (sessionId: Int, fileType: Int, promise: Promise) in
      guard Self.fileExtension(for: fileType) != nil else {
        promise.reject(UteException("BadFileType: unknown file type \(fileType)."))
        return
      }
      self.bridge.deleteFile(sessionId: sessionId, fileType: fileType) { errorCode, result in
        guard errorCode == 0, result == 0 else {
          let why = [1: "it is recording", 2: "the file is starred", 3: "it is playing", 4: "it is in USB mode"][result]
          promise.reject(UteException("DeleteFailed: the clip can't delete now\(why.map { " — \($0)" } ?? "") (error \(errorCode), result \(result))."))
          return
        }
        promise.resolve(true)
      }
    }
  }

  private static func settle(_ promise: Promise, _ errorCode: Int, _ result: [String: Any]?, _ call: String) {
    guard errorCode == 0, let result else {
      promise.reject(UteException("\(call) failed: \(describe(errorCode))."))
      return
    }
    promise.resolve(result)
  }

  /// UTEDeviceError values that can actually come back from the record calls.
  private static func describe(_ code: Int) -> String {
    switch code {
    case -1000: return "the clip is not connected"
    case -999: return "the reply was corrupted (CRC)"
    case -996: return "the clip's Bluetooth service isn't ready yet — wait a moment after connecting"
    case -601: return "the clip sent an empty reply"
    case -600: return "this firmware doesn't support that command"
    case -2: return "the clip refused"
    case 408: return "the clip didn't answer in time"
    default: return "error \(code)"
    }
  }

  private func finishSync(sessionId: Int, data: Data, error: String?) {
    guard let promise = syncPromise else { return }
    syncPromise = nil
    if let error {
      promise.reject(UteException("SyncFailed: \(error)."))
      return
    }
    let ext = syncFileExtension

    // Decoding a long recording takes a moment; keep it off the main thread.
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        let dir = try Self.recordingsDirectory()
        let raw = dir.appendingPathComponent("\(sessionId).\(ext)")
        try data.write(to: raw, options: .atomic)

        var result: [String: Any] = [
          "sessionId": sessionId,
          "path": raw.path,
          "uri": raw.absoluteString,
          "bytes": data.count,
        ]
        do {
          result.merge(try Self.decodeToWav(data, next: raw)) { _, new in new }
        } catch {
          // Keep the raw file; the app can still show it and retry decoding later.
          result["decodeError"] = error.localizedDescription
        }
        promise.resolve(result)
      } catch {
        promise.reject(UteException("WriteFailed: \(error.localizedDescription)"))
      }
    }
  }

  private static func decodeToWav(_ data: Data, next source: URL) throws -> [String: Any] {
    let wav = source.deletingPathExtension().appendingPathExtension("wav")
    let decoded = try OpusWav.decode(data, to: wav)
    return [
      "wavPath": wav.path,
      "wavUri": wav.absoluteString,
      "seconds": decoded.seconds,
      "channels": decoded.channels,
      "sampleRate": decoded.sampleRate,
      "badPackets": decoded.badPackets,
    ]
  }

  /// Documents/recordings — the same place the app's Record tab keeps its list.
  private static func recordingsDirectory() throws -> URL {
    let dir = try FileManager.default
      .url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
      .appendingPathComponent("recordings", isDirectory: true)
    try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    return dir
  }

  /// UTERecordingFileType raw values from UTERecordMgr.h.
  private static func fileExtension(for fileType: Int) -> String? {
    switch fileType {
    case 0: return "avo"
    case 1: return "opus"
    case 2: return "mp3"
    case 3: return "sbc"
    case 4: return "pcm"
    case 5: return "wav"
    default: return nil
    }
  }
}

File: jarvis/app/modules/ute-ble/android/src/main/java/expo/modules/uteble/UteBleModule.kt
package expo.modules.uteble

import android.bluetooth.BluetoothGattCharacteristic
import android.content.Context
import android.util.Base64
import com.yc.nadalsdk.ble.open.UteBleClient
import com.yc.nadalsdk.ble.open.UteBleConnection
import com.yc.nadalsdk.ble.open.UteBleDevice
import com.yc.nadalsdk.bean.DevicePairedState
import com.yc.nadalsdk.bean.HonorAccountConfig
import com.yc.nadalsdk.bean.Notify
import com.yc.nadalsdk.bean.RegionLockConfig
import com.yc.nadalsdk.bean.recorder.AudioRecordStopInfo
import com.yc.nadalsdk.bean.recorder.DeviceStorageInfo
import com.yc.nadalsdk.bean.recorder.RequestAudioRecordFileInfo
import com.yc.nadalsdk.bean.recorder.RequestDeleteAudioRecordFileInfo
import com.yc.nadalsdk.bean.recorder.RequestSyncAudioRecordFileInfo
import com.yc.nadalsdk.bean.recorder.SyncAudioDataInfo
import com.yc.nadalsdk.constants.NotifyType
import com.yc.nadalsdk.constants.recorder.DeleteFileResult
import com.yc.nadalsdk.constants.recorder.FileStatusConstants
import com.yc.nadalsdk.listener.BleConnectStateListener
import com.yc.nadalsdk.listener.ClipDeviceWriteListener
import com.yc.nadalsdk.listener.DeviceNotifyListener
import com.yc.nadalsdk.scan.UteScanCallback
import com.yc.nadalsdk.scan.UteScanDevice
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.util.UUID
import java.util.concurrent.Executors

// Same JS surface as ios/src/UteBleModule.swift. The Android SDK differs in
// three ways that shape this file: its calls are synchronous (so they must stay
// off the main thread), session ids are strings, and the device expects the
// app to answer a pairing notify and a region-lock request after connecting.

private class NotInitialized : CodedException("Call initialize() before anything else.")

private class SdkFailed(call: String, code: Int) : CodedException("$call returned error $code")

class UteBleModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private var client: UteBleClient? = null

  private val connection: UteBleConnection
    get() = client?.uteBleConnection ?: throw NotInitialized()

  // Replies to device-initiated requests arrive on the SDK's callback thread;
  // calling back into the synchronous SDK from there risks blocking it.
  private val replies = Executors.newSingleThreadExecutor()

  private var syncPromise: Promise? = null
  private var syncSessionId = 0L
  private var syncTotal = 0L
  private var syncExtension = "opus"

  // The SDK has no status query on Android, so this tracks what the app did.
  private var recording = false

  override fun definition() = ModuleDefinition {
    Name("UteBle")

    Events(
      "onDeviceFound",
      "onConnectionChange",
      "onPairingChange",
      "onBluetoothState",
      "onRecordStart",
      "onRecordStop",
      "onSyncProgress",
      "onClipData"
    )

    AsyncFunction("initialize") {
      val c = UteBleClient.initialize(context.applicationContext)
      // Matches the vendor demo's Application setup.
      c.setSupportUserIdPair(true)
      // Listeners have to be in place before connect() is ever called.
      c.uteBleConnection.setConnectStateListener(connectStateListener)
      c.uteBleConnection.setDeviceNotifyListener(notifyListener)
      c.uteBleConnection.setClipDeviceWriteListener(clipListener)
      client = c
      sendEvent("onBluetoothState", mapOf("state" to if (c.isBluetoothEnable) 0 else 1, "poweredOn" to c.isBluetoothEnable))
      "1.3.5-android"
    }

    AsyncFunction("startScan") {
      val c = client ?: throw NotInitialized()
      c.cancelScan()
      val started = c.scanDevice(object : UteScanCallback {
        override fun onScanning(device: UteScanDevice) = emitDevice(device)
        override fun onScanComplete(devices: MutableList<UteScanDevice>) = Unit
        override fun onScanFailed(errorCode: Int) {
          sendEvent("onBluetoothState", mapOf("state" to errorCode, "poweredOn" to c.isBluetoothEnable))
        }
      }, 10_000L)
      if (!started) throw CodedException("Scan didn't start. Is Bluetooth on and permission granted?")
    }

    AsyncFunction("stopScan") {
      client?.cancelScan()
    }

    // Android connects by MAC address directly; no scan result object needed.
    AsyncFunction("connect") { id: String ->
      val c = client ?: throw NotInitialized()
      c.cancelScan()
      c.connect(id)
      Unit
    }

    AsyncFunction("disconnect") {
      val c = client ?: return@AsyncFunction false
      val wasConnected = c.isConnected
      c.disconnect()
      wasConnected
    }

    AsyncFunction("isConnected") {
      client?.isConnected ?: false
    }

    // The Android SDK exposes no firmware capability flags, so the recording
    // support fields are null (unknown) rather than guessed.
    AsyncFunction("connectedDevice") {
      val c = client ?: return@AsyncFunction null
      if (!c.isConnected) return@AsyncFunction null
      mapOf(
        "name" to (c.deviceName ?: ""),
        "address" to (c.deviceAddress ?: ""),
        "model" to "",
        "firmware" to "",
        "serialNumber" to "",
        "hasAIRecording" to null,
        "hasAIRecordRealTime" to null,
      )
    }

    // There is no recorder bind on Android; the demo's "request pairing" is the
    // equivalent handshake. token/verify are iOS-only and ignored here.
    AsyncFunction("bind") { _: String, _: Int ->
      val response = connection.requestDevicePairing(1)
      if (!response.isSuccess) throw SdkFailed("requestDevicePairing", response.errorCode)
      val bound = response.data?.pairedState == DevicePairedState.STATE_CONNECTED
      mapOf("ssn" to null, "bound" to bound, "refuseCause" to null, "micMode" to null)
    }

    AsyncFunction("getStatus") {
      mapOf(
        "state" to 0, // Unknown: no status query exists in the Android SDK.
        "recording" to recording,
        "usbConnected" to null,
        "privacyMode" to null,
      )
    }

    AsyncFunction("getStorageInfo") {
      val response = connection.deviceStorageInfo
      val info = response.data
      if (!response.isSuccess || info == null) throw SdkFailed("getDeviceStorageInfo", response.errorCode)
      mapOf(
        "totalKB" to info.total,
        "freeKB" to info.free,
        "bytesPerSecond" to info.recSizePs,
        "full" to (info.freeSizeState == DeviceStorageInfo.NO_FREE_SPACE),
      )
    }

    AsyncFunction("startRecord") {
      val response = connection.appStartAudioRecord()
      val info = response.data
      if (!response.isSuccess || info == null) throw SdkFailed("appStartAudioRecord", response.errorCode)
      recording = info.result == 0
      mapOf("sessionId" to sessionNumber(info.sessionId), "result" to info.result)
    }

    AsyncFunction("stopRecord") {
      val response = connection.appStopAudioRecord()
      val info = response.data
      if (!response.isSuccess || info == null) throw SdkFailed("appStopAudioRecord", response.errorCode)
      recording = false
      mapOf(
        "sessionId" to sessionNumber(info.sessionId),
        "saved" to (info.fileExist == AudioRecordStopInfo.SAVED),
        "fileSize" to info.fileSize,
      )
    }

    // uid 1 / start 0 / only_one 0: the iOS docs say "fill with 0"; the Android
    // demo's only_one = 1 looks like it limits the reply to a single file.
    AsyncFunction("listFiles") {
      val response = connection.queryAudioRecordFileLists(RequestAudioRecordFileInfo(1L, 0L, 0L))
      val info = response.data
      if (!response.isSuccess || info == null) throw SdkFailed("queryAudioRecordFileLists", response.errorCode)
      val files = (info.audioRecordFiles ?: emptyList()).map { file ->
        mapOf(
          "sessionId" to sessionNumber(file.sessionId),
          "size" to file.fileSize,
          "type" to file.fileType,
        )
      }
      mapOf("count" to info.count, "files" to files)
    }

    // One request for the whole file. The SDK accumulates the packets itself
    // and hands over the complete file in AI_RECORDER_SYNC_ALL_DATA_NOTIFY;
    // per-packet progress arrives as AI_RECORDER_SYNCING_ALL_DATA_NOTIFY.
    AsyncFunction("syncFile") { sessionId: Long, fileType: Int, size: Long, promise: Promise ->
      if (syncPromise != null) {
        promise.reject("SyncBusy", "A sync is already running.", null)
        return@AsyncFunction
      }
      syncPromise = promise
      syncSessionId = sessionId
      syncTotal = size
      syncExtension = fileExtension(fileType)

      val request = RequestSyncAudioRecordFileInfo().apply {
        setSessionId(sessionId.toString())
        setFileType(fileType)
        setStart(0L)
        setEnd(size)
        setRealSync(false)
      }
      val response = try {
        connection.syncAudioRecordFile(request)
      } catch (e: Exception) {
        failSync("SyncFailed", e.message ?: "syncAudioRecordFile threw")
        return@AsyncFunction
      }
      val status = response.data?.fileStatus
      if (!response.isSuccess || status != FileStatusConstants.FILE_STATUS_OK) {
        // 1 filesystem error, 2 missing, 3 interrupted. No data will follow.
        failSync("SyncRefused", "syncAudioRecordFile error=${response.errorCode} status=$status")
      }
    }

    AsyncFunction("stopSync") {
      connection.stopSyncAudioRecordFile()
      failSync("SyncStopped", "Sync cancelled.")
    }

    AsyncFunction("deleteFile") { sessionId: Long, fileType: Int ->
      val request = RequestDeleteAudioRecordFileInfo().apply {
        setSessionId(sessionId.toString())
        setFileType(fileType)
      }
      val response = connection.deleteAudioRecordFile(request)
      val result = response.data?.result
      if (!response.isSuccess || result != DeleteFileResult.DELETE_SUCCESS) {
        throw CodedException("deleteAudioRecordFile error=${response.errorCode} result=$result")
      }
      true
    }

    OnDestroy {
      replies.shutdown()
    }
  }

  // MARK: - Listeners

  private val connectStateListener = BleConnectStateListener { status ->
    when (status) {
      // Codes normalised to the iOS UTEDevicesStatus values the JS types document.
      BleConnectStateListener.STATE_CONNECTED -> emitConnection(0, true)
      BleConnectStateListener.STATE_DISCONNECTED -> {
        recording = false
        emitConnection(1, false)
      }
      BleConnectStateListener.STATE_CONNECTING -> emitConnection(4, false)
      // The device asks for a region right after connecting; the demo answers
      // "inside region" every time.
      BleConnectStateListener.STATE_SET_REGION_LOCK -> replies.execute {
        runCatching { connection.setRegionLockConfig(RegionLockConfig(RegionLockConfig.INSIDE_REGION)) }
      }
    }
  }

  private val notifyListener = DeviceNotifyListener { _: UteBleDevice, notify: Notify ->
    when (notify.type) {
      NotifyType.DEVICE_PAIRED_STATE_NOTIFY -> onPairedState(notify.data as? DevicePairedState)

      NotifyType.AI_RECORDER_SYNCING_ALL_DATA_NOTIFY -> {
        val data = notify.data as? SyncAudioDataInfo ?: return@DeviceNotifyListener
        sendEvent("onSyncProgress", mapOf(
          "sessionId" to syncSessionId,
          "received" to data.position,
          "total" to syncTotal,
          "completed" to false,
        ))
      }

      NotifyType.AI_RECORDER_SYNC_ALL_DATA_NOTIFY -> {
        val data = notify.data as? SyncAudioDataInfo ?: return@DeviceNotifyListener
        finishSync(data.syncAudioData ?: ByteArray(0))
      }

      NotifyType.AI_RECORDER_ABORT_SYNC_DATA_NOTIFY -> {
        val received = (notify.data as? SyncAudioDataInfo)?.syncAudioData?.size ?: 0
        failSync("SyncAborted", "The device aborted the transfer after $received bytes.")
      }
    }
  }

  private val clipListener = object : ClipDeviceWriteListener {
    override fun onClipDeviceWrite(characteristic: BluetoothGattCharacteristic, status: Int) = Unit

    @Suppress("DEPRECATION") // getValue() is the only accessor on the SDK's callback.
    override fun onClipDeviceNotify(characteristic: BluetoothGattCharacteristic) {
      val bytes = characteristic.value ?: return
      sendEvent("onClipData", mapOf(
        "bytes" to bytes.size,
        "base64" to Base64.encodeToString(bytes, Base64.NO_WRAP),
      ))
    }
  }

  /**
   * The demo's handshake: when the device reports it is paired, it expects the
   * app to send a stable user id; anything else and the demo disconnects.
   */
  private fun onPairedState(state: DevicePairedState?) {
    val paired = state?.pairedState == DevicePairedState.STATE_CONNECTED
    sendEvent("onPairingChange", mapOf("paired" to paired))
    replies.execute {
      runCatching {
        if (paired) {
          connection.setHonorAccount(HonorAccountConfig().apply { currentHuid = userId() })
        } else {
          client?.disconnect()
        }
      }
    }
  }

  // MARK: - Helpers

  private fun emitDevice(device: UteScanDevice) {
    val bt = device.device ?: return
    // Reading the name throws without BLUETOOTH_CONNECT on Android 12+.
    val name = runCatching { bt.name }.getOrNull()
    if (name.isNullOrEmpty()) return // the demo drops unnamed devices too

    val record = device.scanRecord ?: ByteArray(0)
    val hex = record.joinToString(" ") { "%02X".format(it) }
    // The demo's own filter: UTE firmware advertises 55 55 (or 3A 55 for
    // glasses); the recording clip's name starts with "Tic".
    val likelyUte = hex.contains("55 55") || hex.contains("3A 55") ||
      name.startsWith("AT") || name.startsWith("Tic")

    sendEvent("onDeviceFound", mapOf(
      "id" to bt.address,
      "name" to name,
      "address" to bt.address,
      "rssi" to device.rssi,
      "likelyUte" to likelyUte,
    ))
  }

  private fun emitConnection(status: Int, connected: Boolean) {
    sendEvent("onConnectionChange", mapOf("status" to status, "connected" to connected, "error" to null))
  }

  private fun finishSync(data: ByteArray) {
    val promise = syncPromise ?: return
    syncPromise = null
    try {
      val dir = File(context.cacheDir, "ute-recordings").apply { mkdirs() }
      val file = File(dir, "$syncSessionId.$syncExtension")
      file.writeBytes(data)
      sendEvent("onSyncProgress", mapOf(
        "sessionId" to syncSessionId,
        "received" to data.size,
        "total" to syncTotal,
        "completed" to true,
      ))
      promise.resolve(mapOf(
        "sessionId" to syncSessionId,
        "path" to file.absolutePath,
        "uri" to "file://${file.absolutePath}",
        "bytes" to data.size,
      ))
    } catch (e: Exception) {
      promise.reject("WriteFailed", e.message, e)
    }
  }

  private fun failSync(code: String, message: String) {
    val promise = syncPromise ?: return
    syncPromise = null
    promise.reject(code, message, null)
  }

  private fun userId(): String {
    val prefs = context.getSharedPreferences("ute-ble", Context.MODE_PRIVATE)
    return prefs.getString("userId", null) ?: UUID.randomUUID().toString().replace("-", "").also {
      prefs.edit().putString("userId", it).apply()
    }
  }

  // Session ids are numeric strings on Android; JS gets them as numbers, the
  // same as iOS.
  private fun sessionNumber(id: String?): Long = id?.toLongOrNull() ?: 0L

  private fun fileExtension(type: Int) = when (type) {
    0 -> "avo"
    1 -> "opus"
    2 -> "mp3"
    3 -> "sbc"
    4 -> "pcm"
    5 -> "wav"
    else -> "bin"
  }
}
